File: http/typescript/fetch/src/middlewares/options/redirectHandlerOption.ts
/**
 * -------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
 * See License in the project root for license information.
 * -------------------------------------------------------------------------------------------
 */

/**
 * @module RedirectHandlerOptions
 */

import { RequestOption } from "@microsoft/kiota-abstractions";

/**
 * @type
 * A type declaration for shouldRetry callback
 */
export type ShouldRedirect = (response: Response) => boolean;

/**
 * @class
 * @implements MiddlewareOptions
 * A class representing RedirectHandlerOptions
 */
export class RedirectHandlerOptions implements RequestOption {
    /**
     * @private
     * @static
     * A member holding default max redirects value
     */
    private static DEFAULT_MAX_REDIRECTS = 5;

    /**
     * @private
     * @static
     * A member holding maximum max redirects value
     */
    private static MAX_MAX_REDIRECTS = 20;

    /**
     * @public
     * A member holding max redirects value
     */
    public maxRedirects: number;

    /**
     * @public
     * A member holding shouldRedirect callback
     */
    public shouldRedirect: ShouldRedirect;

    /**
     * @private
     * A member holding default shouldRedirect callback
     */
    private static defaultShouldRetry: ShouldRedirect = () => true;

    /**
     * @public
     * @constructor
     * To create an instance of RedirectHandlerOptions
     * @param {number} [maxRedirects = RedirectHandlerOptions.DEFAULT_MAX_REDIRECTS] - The max redirects value
     * @param {ShouldRedirect} [shouldRedirect = RedirectHandlerOptions.DEFAULT_SHOULD_RETRY] - The should redirect callback
     * @returns An instance of RedirectHandlerOptions
     */
    public constructor(maxRedirects: number = RedirectHandlerOptions.DEFAULT_MAX_REDIRECTS, shouldRedirect: ShouldRedirect = RedirectHandlerOptions.defaultShouldRetry) {
        if (maxRedirects > RedirectHandlerOptions.MAX_MAX_REDIRECTS) {
            const error = new Error(`MaxRedirects should not be more than ${RedirectHandlerOptions.MAX_MAX_REDIRECTS}`);
            error.name = "MaxLimitExceeded";
            throw error;
        }
        if (maxRedirects < 0) {
            const error = new Error(`MaxRedirects should not be negative`);
            error.name = "MinExpectationNotMet";
            throw error;
        }
        this.maxRedirects = maxRedirects;
        this.shouldRedirect = shouldRedirect;
    }

    public getKey():string {
    // TODO
    return "";
    }
}

File: http/java/okhttp/lib/src/main/java/com/microsoft/kiota/http/middleware/options/RedirectHandlerOption.java
package com.microsoft.kiota.http.middleware.options;

import javax.annotation.Nullable;

import com.microsoft.kiota.RequestOption;

import javax.annotation.Nonnull;

/**
 * Options to be passed to the redirect middleware.
 */
public class RedirectHandlerOption implements RequestOption {
    private int maxRedirects;
    /**
     * The default maximum number of redirects to follow
     */
    public static final int DEFAULT_MAX_REDIRECTS = 5;
    /**
     * The absolute maxium number of redirects that can be followed
     */
    public static final int MAX_REDIRECTS = 20;

    private IShouldRedirect shouldRedirect;
    /**
     * Default redirect evaluation, always follow redirect information.
     */
    public static final IShouldRedirect DEFAULT_SHOULD_REDIRECT = response -> true;

    /**
     * Create default instance of redirect options, with default values of max redirects and should redirect
     */
    public RedirectHandlerOption() {
        this(DEFAULT_MAX_REDIRECTS, DEFAULT_SHOULD_REDIRECT);
    }

    /**
     * @param maxRedirects Max redirects to occur
     * @param shouldRedirect Should redirect callback called before every redirect
     */
    public RedirectHandlerOption(int maxRedirects, @Nullable final IShouldRedirect shouldRedirect) {
        if(maxRedirects < 0)
            throw new IllegalArgumentException("Max redirects cannot be negative");
        if(maxRedirects > MAX_REDIRECTS)
            throw new IllegalArgumentException("Max redirect cannot exceed " + MAX_REDIRECTS);

        this.maxRedirects = maxRedirects;
        this.shouldRedirect = shouldRedirect != null ? shouldRedirect : DEFAULT_SHOULD_REDIRECT;
    }

    /**
     * @return max redirects
     */
    public int maxRedirects() {
        return this.maxRedirects;
    }

    /**
     * @return should redirect
     */
    @Nonnull
    public IShouldRedirect shouldRedirect() {
        return this.shouldRedirect;
    }

    @Override
    public  Class<? extends RequestOption> getType() {
        return RedirectHandlerOption.class; 
    }
}
